pl_examples/loops_customisation/k_fold.py

```diff
@@ -0,0 +1,162 @@
+        return self.dataset[self.indices[index]]
+
+    def __len__(self) -> int:
+        return len(self.indices) - 1
+
+
+@dataclass
+class KFoldLoop(ExternalLoop):
+
+    num_folds: int
+    num_epochs: int = 10
+    best_model_paths: List[str] = field(default_factory=lambda: [])
+    restarting: bool = False
+
+    @staticmethod
+    def loop_base_callback() -> Type[Callback]:
+        class BaseKFoldCallback(Callback):
+            @rank_zero_only
+            def on_fold_start(self, trainer, pl_module, counter):
+                """Override with your own logic"""
+
+        return BaseKFoldCallback
+
+    @property
+    def done(self) -> bool:
+        return self.current_fold >= self.num_folds
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.current_fold = 0
+            self.set_max_epochs(self.num_epochs)
+
+    def generate_fold(self, dataloader_kwargs: Dict[str, Any], stage: str):
+        dataset = dataloader_kwargs["dataset"]
+        kfold = KFold(self.num_folds, random_state=42, shuffle=True)
+        train_indices, validation_indices = list(kfold.split(range(len(dataset))))[self.current_fold]
+        if stage == "train":
+            dataloader_kwargs["dataset"] = SplitDataset(dataset, train_indices.tolist())
+        else:
+            dataloader_kwargs["dataset"] = SplitDataset(dataset, validation_indices.tolist())
+        dataloader_kwargs["sampler"].data_source = dataloader_kwargs["dataset"]
+        return dataloader_kwargs
+
+    def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        # temporary hack
+        self.trainer.datamodule.setup("fit")
+
+    def on_advance_start(self):
+        self.reload_train_dataloader(self.generate_fold)
+        self.reload_val_dataloaders(self.generate_fold)
+        self.trainer.call_hook("on_fold_start", self.current_fold)
+        self.lightning_module.reset_parameters()
+
+    def advance(self):
+        return self.trainer.fit(self.lightning_module, train_dataloader=self.train_dataloader)
+
+    def on_advance_end(self) -> None:
+        self.current_fold += 1
+        self.increment_max_epochs(self.num_epochs)
+        # stored best weight path for this fold
+        self.best_model_paths.append(self.trainer.checkpoint_callback.best_model_path)
+
+    def on_save_checkpoint(self) -> Dict:
+        return {"current_fold": self.current_fold}
+
+    def on_load_checkpoint(self, state_dict) -> None:
+        self.current_fold = state_dict["current_fold"]
+
+
+class KFoldCallback(KFoldLoop.loop_base_callback()):
+
+    """This callback demonstrates how to implement your own callback API."""
+
+    @rank_zero_only
+    def on_fold_start(self, trainer, pl_module, counter):
+        log.info(f"Starting to train on fold {counter}")
+
+
+loop = KFoldLoop(5)
+model = BoringModel()
+datamodule = BoringDataModule()
+trainer = Trainer(callbacks=KFoldCallback())
+trainer.run_loop(model, datamodule=datamodule, loop=loop)
```

Comment:
I don't think we should introduce something like this. IMO the w should be the atomic element here and not call something that iterates on it. This should ratjer be implemented as a standalone utility class (as discussed in the corresponding issue (can't fond the number right now))

Reply:
I guess it is just remaining after debugging